File: app/src/main/java/io/github/fvasco/pinpoi/dao/PlacemarkDao.java
package io.github.fvasco.pinpoi.dao;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteOpenHelper;
import android.os.Build;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.SortedSet;
import java.util.TreeSet;

import io.github.fvasco.pinpoi.model.Placemark;
import io.github.fvasco.pinpoi.model.PlacemarkAnnotation;
import io.github.fvasco.pinpoi.model.PlacemarkBase;
import io.github.fvasco.pinpoi.model.PlacemarkSearchResult;
import io.github.fvasco.pinpoi.util.Coordinates;
import io.github.fvasco.pinpoi.util.DistanceComparator;
import io.github.fvasco.pinpoi.util.Util;

/**
 * Dao for {@linkplain io.github.fvasco.pinpoi.model.Placemark}
 *
 * @author Francesco Vasco
 */
/*
 * Save coordinate as int: coordinate*{@linkplain #COORDINATE_MULTIPLIER}
 */
public class PlacemarkDao extends AbstractDao<PlacemarkDao> {

    /**
     * Max result for {@linkplain #findAllPlacemarkNear(Coordinates, double, String, boolean, Collection)}
     */
    private static final int MAX_NEAR_RESULT = 250;
    private static final boolean SQL_INSTR_PRESENT = Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP;

    // 2^20
    private static final float COORDINATE_MULTIPLIER = 1048576F;
    private static PlacemarkDao INSTANCE;

    PlacemarkDao() {
        this(Util.getApplicationContext());
    }

    public PlacemarkDao(@NonNull Context context) {
        super(context);
    }

    public static synchronized PlacemarkDao getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new PlacemarkDao();
        }
        return INSTANCE;
    }

    /**
     * Convert DB coordinates to double
     *
     * @param i db coordinate
     * @return float coordinates
     */
    private static float coordinateToFloat(int i) {
        return i / COORDINATE_MULTIPLIER;
    }

    /**
     * Convert double to db coordinates
     *
     * @param f float coordinate
     * @return db coordinates
     */
    private static int coordinateToInt(float f) {
        return Math.round(f * COORDINATE_MULTIPLIER);
    }

    private static int coordinateToInt(double f) {
        return (int) Math.round(f * COORDINATE_MULTIPLIER);
    }

    /**
     * Append coordinates filter in stringBuilder sql clause
     */
    private static void createWhereFilter(Coordinates coordinates, double range, final String table, StringBuilder stringBuilder) {
        // calculate "square" of search
        final Coordinates shiftY = coordinates.withLatitude(coordinates.getLatitude() + (coordinates.getLatitude() > 0 ? -1 : 1));
        final float scaleY = coordinates.distanceTo(shiftY);
        final Coordinates shiftX = coordinates.withLongitude(coordinates.getLongitude() + (coordinates.getLongitude() > 0 ? -1 : 1));
        final float scaleX = coordinates.distanceTo(shiftX);

        // latitude
        stringBuilder.append(table).append(".latitude between ")
                .append(String.valueOf(coordinateToInt(coordinates.getLatitude() - range / scaleY))).append(" AND ")
                .append(String.valueOf(coordinateToInt(coordinates.getLatitude() + range / scaleY)));

        // longitude
        final double longitudeMin = coordinates.getLongitude() - range / scaleX;
        final double longitudeMax = coordinates.getLongitude() + range / scaleX;
        stringBuilder.append(" AND (").append(table).append(".longitude between ")
                .append(String.valueOf(coordinateToInt(longitudeMin))).append(" AND ")
                .append(String.valueOf(coordinateToInt(longitudeMax)));
        // fix for meridian 180
        if (longitudeMin < -180.0) {
            stringBuilder.append(" OR ").append(table).append(".longitude >=").append(String.valueOf(coordinateToInt(longitudeMin + 360.0)));
        } else if (longitudeMax > 180.0) {
            stringBuilder.append(" OR ").append(table).append(".longitude <=").append(String.valueOf(coordinateToInt(longitudeMax - 360.0)));
        }
        stringBuilder.append(')');
    }

    @Override
    protected SQLiteOpenHelper createSqLiteOpenHelper(@NonNull Context context) {
        return new PlacemarkDatabase(context);
    }

    public List<Placemark> findAllPlacemarkByCollectionId(final long collectionId) {
        final Cursor cursor = database.query("PLACEMARK", null,
                "collection_id=" + collectionId, null, null, null, "_ID");
        try {
            final List<Placemark> res = new ArrayList<>();
            cursor.moveToFirst();
            while (!cursor.isAfterLast()) {
                res.add(cursorToPlacemark(cursor));
                cursor.moveToNext();
            }
            return res;
        } finally {
            cursor.close();
        }
    }

    public SortedSet<PlacemarkSearchResult> findAllPlacemarkNear
            (final Coordinates coordinates,
             final double range,
             final Collection<Long> collectionIds) {
        return findAllPlacemarkNear(coordinates, range, null, false, collectionIds);
    }

    /**
     * Search {@linkplain Placemark} near location
     *
     * @param coordinates   the center of search
     * @param range         radius of search, in meters
     * @param collectionIds collection id filter
     */
    public SortedSet<PlacemarkSearchResult> findAllPlacemarkNear(
            final Coordinates coordinates,
            final double range,
            String nameFilter,
            final boolean onlyFavourite,
            final Collection<Long> collectionIds) {
        Util.requireNonNull(coordinates, "coordinates not set");
        Util.requireNonNull(collectionIds, "collection not set");
        if (collectionIds.isEmpty()) {
            throw new IllegalArgumentException("collection empty");
        }
        if (range <= 0) {
            throw new IllegalArgumentException("range not valid " + range);
        }
        // nameFilter null or UPPERCASE
        if (Util.isEmpty(nameFilter)) {
            nameFilter = null;
        } else {
            nameFilter = nameFilter.trim().toUpperCase();
        }

        // sql clause
        // collection ids
        final StringBuilder sql = new StringBuilder(
                "SELECT p._ID,p.latitude,p.longitude,p.name,pa.flag FROM PLACEMARK p")
                .append(" LEFT OUTER JOIN PLACEMARK_ANNOTATION pa USING(latitude,longitude)");
        sql.append(" WHERE p.collection_id in (");
        final List<String> whereArgs = new ArrayList<>();
        final Iterator<Long> iterator = collectionIds.iterator();
        sql.append(iterator.next().toString());
        while (iterator.hasNext()) {
            sql.append(',').append(iterator.next().toString());
        }
        sql.append(") AND ");
        createWhereFilter(coordinates, range, "p", sql);

        if (onlyFavourite) {
            sql.append(" AND pa.flag=1");
        }

        if (SQL_INSTR_PRESENT && nameFilter != null) {
            sql.append(" AND instr(upper(name),?)>0");
            whereArgs.add(nameFilter);
        }

        final DistanceComparator locationComparator = new DistanceComparator(coordinates);
        final SortedSet<PlacemarkSearchResult> res = new TreeSet<>(locationComparator);
        final Cursor cursor = database.rawQuery(sql.toString(), whereArgs.toArray(new String[whereArgs.size()]));
        try {
            cursor.moveToFirst();
            double maxDistance = range;
            while (!cursor.isAfterLast()) {
                final PlacemarkSearchResult p = cursorToPlacemarkSearchResult(cursor);
                if (locationComparator.calculateDistance(p) <= maxDistance
                        && (SQL_INSTR_PRESENT || nameFilter == null || p.getName().toUpperCase().contains(nameFilter))) {
                    res.add(p);
                    // ensure size limit, discard farest
                    if (res.size() > MAX_NEAR_RESULT) {
                        final PlacemarkSearchResult placemarkToDiscard = res.last();
                        res.remove(placemarkToDiscard);
                        // update search range to search closer
                        maxDistance = locationComparator.calculateDistance(res.last());
                    }
                }
                cursor.moveToNext();
            }
        } finally {
            cursor.close();
        }
        return res;
    }

    public Placemark getPlacemark(final long id) {
        final Cursor cursor = database.query("PLACEMARK", null,
                "_ID=" + id, null, null, null, null);
        try {
            cursor.moveToFirst();
            return cursor.isAfterLast() ? null : cursorToPlacemark(cursor);
        } finally {
            cursor.close();
        }
    }

    /**
     * Get annotation for a placemark
     *
     * @return annotaion for a placemark
     */
    public PlacemarkAnnotation loadPlacemarkAnnotation(PlacemarkBase placemark) {
        final Cursor cursor = database.query("PLACEMARK_ANNOTATION", null,
                "latitude=" + coordinateToInt(placemark.getLatitude()) + " AND longitude=" + coordinateToInt(placemark.getLongitude()), null,
                null, null, null);
        try {
            PlacemarkAnnotation res = null;
            cursor.moveToFirst();
            if (!cursor.isAfterLast()) {
                res = cursorToPlacemarkAnnotation(cursor);
            }
            cursor.close();

            if (res == null) {
                res = new PlacemarkAnnotation();
                res.setLatitude(placemark.getLatitude());
                res.setLongitude(placemark.getLongitude());
                res.setNote("");
            }
            return res;
        } finally {
            cursor.close();
        }
    }

    public void update(final PlacemarkAnnotation placemarkAnnotation) {
        if (placemarkAnnotation.getNote().isEmpty() && !placemarkAnnotation.isFlagged()) {
            database.delete("PLACEMARK_ANNOTATION", "_ID=" + placemarkAnnotation.getId(), null);
            placemarkAnnotation.setId(0);
        } else {
            if (placemarkAnnotation.getId() > 0) {
                final int count = database.update("PLACEMARK_ANNOTATION", placemarkAnnotationToContentValues(placemarkAnnotation), "_ID=" + placemarkAnnotation.getId(), null);
                if (count == 0) {
                    placemarkAnnotation.setId(0);
                }
            }
            if (placemarkAnnotation.getId() == 0) {
                final long id = database.insert("PLACEMARK_ANNOTATION", null, placemarkAnnotationToContentValues(placemarkAnnotation));
                if (id == -1) {
                    throw new IllegalArgumentException("Data not valid");
                }
                placemarkAnnotation.setId(id);
            }
        }
    }

    public void insert(Placemark p) {
        final long id = database.insert("PLACEMARK", null, placemarkToContentValues(p));
        if (id == -1) {
            throw new IllegalArgumentException("Data not valid");
        }
        p.setId(id);
    }

    public void deleteByCollectionId(final long collectionId) {
        database.delete("PLACEMARK", "collection_id=" + collectionId, null);
    }

    private ContentValues placemarkToContentValues(final Placemark p) {
        final ContentValues cv = new ContentValues();
        cv.put("latitude", coordinateToInt(p.getLatitude()));
        cv.put("longitude", coordinateToInt(p.getLongitude()));
        cv.put("name", p.getName());
        cv.put("description", p.getDescription());
        cv.put("collection_id", p.getCollectionId());
        return cv;
    }

    private ContentValues placemarkAnnotationToContentValues(final PlacemarkAnnotation pa) {
        final ContentValues cv = new ContentValues();
        cv.put("latitude", coordinateToInt(pa.getLatitude()));
        cv.put("longitude", coordinateToInt(pa.getLongitude()));
        cv.put("note", pa.getNote());
        cv.put("flag", pa.isFlagged() ? 1 : 0);
        return cv;
    }

    private Placemark cursorToPlacemark(Cursor cursor) {
        final Placemark p = new Placemark();
        p.setId(cursor.getLong(0));
        p.setLatitude(coordinateToFloat(cursor.getInt(1)));
        p.setLongitude(coordinateToFloat(cursor.getInt(2)));
        p.setName(cursor.getString(3));
        p.setDescription(cursor.getString(4));
        p.setCollectionId(cursor.getLong(5));
        return p;
    }

    private PlacemarkSearchResult cursorToPlacemarkSearchResult(Cursor cursor) {
        return new PlacemarkSearchResult(cursor.getLong(0),
                coordinateToFloat(cursor.getInt(1)),
                coordinateToFloat(cursor.getInt(2)),
                cursor.getString(3),
                cursor.getInt(4) != 0);
    }

    private PlacemarkAnnotation cursorToPlacemarkAnnotation(Cursor cursor) {
        final PlacemarkAnnotation pa = new PlacemarkAnnotation();
        pa.setId(cursor.getLong(0));
        pa.setLatitude(coordinateToFloat(cursor.getInt(1)));
        pa.setLongitude(coordinateToFloat(cursor.getInt(2)));
        pa.setNote(cursor.getString(3));
        pa.setFlagged(cursor.getInt(4) != 0);
        return pa;
    }
}

File: app/src/main/java/io/github/fvasco/pinpoi/dao/PlacemarkDao.kt
package io.github.fvasco.pinpoi.dao

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import io.github.fvasco.pinpoi.model.Placemark
import io.github.fvasco.pinpoi.model.PlacemarkAnnotation
import io.github.fvasco.pinpoi.model.PlacemarkBase
import io.github.fvasco.pinpoi.model.PlacemarkSearchResult
import io.github.fvasco.pinpoi.util.Coordinates
import io.github.fvasco.pinpoi.util.PlacemarkDistanceComparator
import io.github.fvasco.pinpoi.util.distanceTo
import java.util.SortedSet
import java.util.TreeSet
import kotlin.math.roundToInt

/**
 * Dao for [io.github.fvasco.pinpoi.model.Placemark]
 *
 * @author Francesco Vasco
 */
/*
 * Save coordinate as int: coordinate*{@linkplain #COORDINATE_MULTIPLIER}
 */
class PlacemarkDao(context: Context) : AbstractDao(context) {

    override fun createSqLiteOpenHelper(context: Context): SQLiteOpenHelper {
        return PlacemarkDatabase(context)
    }

    fun isEmpty(): Boolean =
        database!!.query("PLACEMARK", arrayOf("_ID"), null, null, null, null, null, "1")
            .use { cursor -> cursor.isAfterLast }

    fun findAllPlacemarkByCollectionId(collectionId: Long): List<Placemark> {
        database!!.query(
            "PLACEMARK", null,
            "collection_id=$collectionId", null, null, null, "_ID"
        ).use { cursor ->
            val res = ArrayList<Placemark>()
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                res.add(cursorToPlacemark(cursor))
                cursor.moveToNext()
            }
            return res
        }
    }

    /**
     * Search [Placemark] near location
     *
     * @param coordinates   the center of search
     * @param range         radius of search, in meters
     * @param collectionIds collection id filter
     */
    fun findAllPlacemarkNear(
        coordinates: Coordinates,
        range: Double,
        collectionIds: Collection<Long>,
        nameFilter: String? = null,
        onlyFavourite: Boolean = false
    ): SortedSet<PlacemarkSearchResult> {
        require(collectionIds.isNotEmpty()) { "collection empty" }
        require(range > 0) { "range not valid $range" }


        // sql clause
        // collection ids
        val sql = StringBuilder(
            "SELECT p._ID,p.latitude,p.longitude,p.name,pa.flag,pa.note,collection_id FROM PLACEMARK p" +
                    " LEFT OUTER JOIN PLACEMARK_ANNOTATION pa USING(latitude,longitude)" +
                    " WHERE p.collection_id in ("
        )
        val whereArgs = ArrayList<String>()
        val iterator = collectionIds.iterator()
        sql.append(iterator.next().toString())
        while (iterator.hasNext()) {
            sql.append(',').append(iterator.next().toString())
        }
        sql.append(") AND ")
        createWhereFilter(coordinates, range, "p", sql)

        if (onlyFavourite) {
            sql.append(" AND pa.flag=1")
        }

        if (!nameFilter.isNullOrBlank()) {
            sql.append(" AND instr(upper(name),?)>0")
            whereArgs.add(nameFilter.uppercase())
        }

        val locationComparator = PlacemarkDistanceComparator(coordinates)
        val res = TreeSet(locationComparator)
        database!!.rawQuery(sql.toString(), whereArgs.toTypedArray()).use { cursor ->
            cursor.moveToFirst()
            var maxDistance = range
            while (!cursor.isAfterLast) {
                val p = cursorToPlacemarkSearchResult(cursor)
                if (coordinates.distanceTo(p.coordinates) <= maxDistance) {
                    res.add(p)
                    // ensure size limit, discard the most distant
                    if (res.size > MAX_NEAR_RESULT) {
                        val placemarkToDiscard = res.last()
                        res.remove(placemarkToDiscard)
                        // update search range to search closer
                        maxDistance = coordinates.distanceTo(res.last().coordinates).toDouble()
                    }
                }
                cursor.moveToNext()
            }
        }
        return res
    }

    fun getPlacemark(id: Long): Placemark? {
        if (id <= 0) return null
        database!!.query(
            "PLACEMARK", null,
            "_ID=$id", null, null, null, null
        ).use { cursor ->
            cursor.moveToFirst()
            return if (cursor.isAfterLast) null else cursorToPlacemark(cursor)
        }
    }

    /**
     * Get annotation for a placemark

     * @return annotaion for a placemark
     */
    fun loadPlacemarkAnnotation(placemark: PlacemarkBase): PlacemarkAnnotation {
        val latitude = placemark.coordinates.latitude
        val longitude = placemark.coordinates.longitude
        database!!.query(
            "PLACEMARK_ANNOTATION",
            null,
            "latitude=" + coordinateToInt(latitude) + " AND longitude=" + coordinateToInt(longitude),
            null,
            null,
            null,
            null
        ).use { cursor ->
            cursor.moveToFirst()
            return if (cursor.isAfterLast) {
                PlacemarkAnnotation(coordinates = Coordinates(latitude, longitude))
            } else {
                cursorToPlacemarkAnnotation(cursor)
            }
        }
    }

    fun update(placemarkAnnotation: PlacemarkAnnotation) {
        val whereClause =
            "latitude=${coordinateToInt(placemarkAnnotation.coordinates.latitude)} AND longitude=${
                coordinateToInt(placemarkAnnotation.coordinates.longitude)
            }"
        if (placemarkAnnotation.note.isBlank() && !placemarkAnnotation.flagged) {
            database!!.delete(
                "PLACEMARK_ANNOTATION",
                whereClause,
                null
            )
        } else {
            val contentValues = placemarkAnnotationToContentValues(placemarkAnnotation)
            val count = database!!.update("PLACEMARK_ANNOTATION", contentValues, whereClause, null)
            if (count == 0) {
                database!!.insertOrThrow("PLACEMARK_ANNOTATION", null, contentValues)
            }
        }
    }

    fun insert(p: Placemark, conflictAlgorithm: Int = SQLiteDatabase.CONFLICT_FAIL): Boolean {
        val id = database!!.insertWithOnConflict(
            "PLACEMARK", null,
            placemarkToContentValues(p), conflictAlgorithm
        )
        return id > 0
    }

    fun deleteByCollectionId(collectionId: Long) {
        database!!.delete("PLACEMARK", "collection_id=$collectionId", null)
    }

    private fun placemarkToContentValues(p: Placemark): ContentValues {
        val c = p.coordinates
        val cv = ContentValues()
        cv.put("latitude", coordinateToInt(c.latitude))
        cv.put("longitude", coordinateToInt(c.longitude))
        cv.put("name", p.name.trim())
        cv.put("description", p.description.trim())
        cv.put("collection_id", p.collectionId)
        return cv
    }

    private fun placemarkAnnotationToContentValues(pa: PlacemarkAnnotation): ContentValues {
        val ca = pa.coordinates
        val cv = ContentValues()
        cv.put("latitude", coordinateToInt(ca.latitude))
        cv.put("longitude", coordinateToInt(ca.longitude))
        cv.put("note", pa.note.trim())
        cv.put("flag", if (pa.flagged) 1 else 0)
        return cv
    }

    private fun cursorToPlacemark(cursor: Cursor) =
        Placemark(
            id = cursor.getLong(0),
            coordinates = Coordinates(
                coordinateToFloat(cursor.getInt(1)),
                coordinateToFloat(cursor.getInt(2))
            ),
            name = cursor.getString(3),
            description = cursor.getString(4),
            collectionId = cursor.getLong(5)
        )

    private fun cursorToPlacemarkSearchResult(cursor: Cursor) =
        PlacemarkSearchResult(
            id = cursor.getLong(0),
            coordinates = Coordinates(
                coordinateToFloat(cursor.getInt(1)),
                coordinateToFloat(cursor.getInt(2))
            ),
            name = cursor.getString(3),
            note = cursor.getString(5)?.takeIf(String::isNotBlank),
            flagged = cursor.getInt(4) != 0,
            collectionId = cursor.getLong(6)
        )

    private fun cursorToPlacemarkAnnotation(cursor: Cursor) =
        PlacemarkAnnotation(
            // column 0 is id (ignored)
            coordinates = Coordinates(
                coordinateToFloat(cursor.getInt(1)),
                coordinateToFloat(cursor.getInt(2))
            ),
            note = cursor.getString(3),
            flagged = cursor.getInt(4) != 0
        )

    companion object {

        /**
         * Max result for [findAllPlacemarkNear]
         */
        private const val MAX_NEAR_RESULT = 300

        // 2^20
        private const val COORDINATE_MULTIPLIER = 1048576f

        /**
         * Convert DB coordinates to double

         * @param i db coordinate
         * *
         * @return float coordinates
         */
        private fun coordinateToFloat(i: Int): Float {
            return i / COORDINATE_MULTIPLIER
        }

        /**
         * Convert double to db coordinates

         * @param f float coordinate
         * *
         * @return db coordinates
         */
        private fun coordinateToInt(f: Float): Int = coordinateToInt(f.toDouble())

        private fun coordinateToInt(f: Double): Int {
            return (f * COORDINATE_MULTIPLIER).roundToInt()
        }

        /**
         * Append coordinates filter in stringBuilder sql clause
         */
        private fun createWhereFilter(
            coordinates: Coordinates,
            range: Double,
            table: String,
            stringBuilder: StringBuilder
        ) {
            // calculate "square" of search
            val shiftY =
                coordinates.copy(latitude = coordinates.latitude + if (coordinates.latitude > 0) -1 else 1)
            val scaleY = coordinates.distanceTo(shiftY)
            val shiftX =
                coordinates.copy(longitude = coordinates.longitude + if (coordinates.longitude > 0) -1 else 1)
            val scaleX = coordinates.distanceTo(shiftX)

            // latitude
            stringBuilder.append(table).append(".latitude between ")
                .append(coordinateToInt(coordinates.latitude - range / scaleY).toString())
                .append(" AND ")
                .append(coordinateToInt(coordinates.latitude + range / scaleY).toString())

            // longitude
            val longitudeMin = coordinates.longitude - range / scaleX
            val longitudeMax = coordinates.longitude + range / scaleX
            stringBuilder.append(" AND (").append(table).append(".longitude between ")
                .append(coordinateToInt(longitudeMin).toString()).append(" AND ")
                .append(coordinateToInt(longitudeMax).toString())
            // fix for meridian 180
            if (longitudeMin < -180.0) {
                stringBuilder.append(" OR ").append(table).append(".longitude >=")
                    .append(coordinateToInt(longitudeMin + 360.0).toString())
            } else if (longitudeMax > 180.0) {
                stringBuilder.append(" OR ").append(table).append(".longitude <=")
                    .append(coordinateToInt(longitudeMax - 360.0).toString())
            }
            stringBuilder.append(')')
        }
    }
}
